src/compiler/config/validate-testing.ts

```diff
@@ -4,20 +4,20 @@ import { isAbsolute, join, basename, dirname } from 'path';
 import { isLocalModule } from '../sys/resolve/resolve-utils';
 import { isOutputTargetDist, isOutputTargetWww } from '../output-targets/output-utils';
 
-export const validateTesting = (config: d.Config, diagnostics: d.Diagnostic[]) => {
+export const validateTesting = (config: d.UnvalidatedConfig, diagnostics: d.Diagnostic[]) => {
   const testing = (config.testing = Object.assign({}, config.testing || {}));
 
   if (!config.flags || (!config.flags.e2e && !config.flags.spec)) {
     return;
   }
 
-  let configPathDir = config.configPath;
+  let configPathDir = config.configPath!;
```

Comment:
For my own understanding, how do we know that `configPath` exists on `config` at this point in the function?

Reply:
Basically we don't (right now) have a type-level assurance that it's not null, but if you look at the `loadConfig` function you'll see that we basically try to read in the configuration file at `configPath` before we call the `validate-config` function with the output, and there is an early return if something goes wrong with reading the config. So basically by the time we're calling `validate-config` (and then subsequently `validate-testing`) we already know that this is present because we've been able to read a file at this path.

This is my understanding of how this works anyway!